### PR TITLE
[Cleanup] batch_norm: reach compute unpack_modes via the arch-agnostic accessor

### LINKED ISSUE #55735: batch_norm: unpack_modes std::get throws on Quasar
### Issue description

`batch_norm` already ported to Metal 2.0 (#51974). Both factories still set compute `unpack_modes` by naming the Gen1 alternative of the variant from `to_compute_hardware_config`:

```cpp
auto& unpack_modes = std::get<ComputeGen1Config>(compute_hw_config).unpack_modes;
```

On Quasar that helper returns `ComputeGen2Config`, so the `std::get` throws `std::bad_variant_access` when `fp32_dest_acc_en` is set.

### Fix

Use the generation-neutral accessor `unpack_modes(config)` — same change as the wording in `moreh_linear_backward`. No `arch` branch. Gen1 behaviour is unchanged.

PR: #55601

Quasar *values* remain #52269. Sibling unowned site in `experimental/quasar/` SDPA is #55733.

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
Issue : https://github.com/tenstorrent/tt-metal/issues/55735

### Summary

Post-port `gen2_hardware_configs` pass on an already-Metal-2.0-ported op. Both `batch_norm` factories reached `unpack_modes` through `std::get<ComputeGen1Config>` on the variant returned by the arch-agnostic `to_compute_hardware_config` — which returns the Gen2 alternative on Quasar, so the `std::get` throws `std::bad_variant_access` there. Replaced with the common-field accessor `unpack_modes(config)`, which resolves whichever alternative the helper chose; no `arch` branch needed.

Nothing changes on Wormhole/Blackhole. On Quasar the old `std::get` throws outright rather than
misbehaving quietly, so it is latent only in the sense that no supported target reaches that path
today. What is deferred is the Quasar *values* (#52269), not the access path.

### Notes for reviewers

- **Gen1 is bit-identical.** Normalizing away the accessor call and the local's name, both files differ from their pre-change versions *only* by the added `TODO` comment — the DFB entry sets are untouched (same names, order, `UnpackToDest`).
- **The `unpack_modes` → `dfb_unpack_modes` rename is forced**, not cosmetic: the old local shadowed the accessor. The spelling matches the idiom documented in `compute_hardware_config.hpp`.
- **Gen1 sentinels green and unchanged** — 1588 passed / 786 xfailed, identical before and after. Nothing here exercises the Gen2 *values*; that branch never runs on a WH/BH bench.

Follow-up: #52269 — the `TODO` in both files points there.
https://github.com/tenstorrent/tt-metal/pull/51974 - BN Porting PR

---

### Pipeline status

- [x] Sanity
- [x] (Runtime) Sanity Test
- [x] Nightly for fused, DM

---

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/norm_post)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/norm_post)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/norm_post)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/norm_post)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/norm_post)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/norm_post)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/norm_post)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp [hunk 1]
```diff
@@ -465,13 +465,14 @@ ttnn::device_operation::ProgramArtifacts BatchNormOperation::BatchNormFactory::c
         // Re-key of the legacy unpack_to_dest_mode vector, which was indexed by CB id. Every DFB the
         // compute kernel consumes is listed; the writer-facing output is producer-only, so it gets no
         // entry. An omitted DFB keeps the UnpackToSrc default.
-        auto& unpack_modes = std::get<ComputeGen1Config>(compute_hw_config).unpack_modes;
+        // TODO(#52269): Quasar unpack_modes are copied from Gen1 and not yet optimized for Quasar.
+        auto& dfb_unpack_modes = unpack_modes(compute_hw_config);
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp:473]: The `TODO(#52269)` carries the *values* half of the rationale but not the *access-path* half. #52269 is about optimizing `unpack_modes` for Quasar; nothing at this site now records that `std::get<ComputeGen1Config>` would throw `std::bad_variant_access` on a Gen2 target, or why the local had to be renamed. Those are exactly the two things a later cleanup pass would undo — it reads as a gratuitous rename plus an unrelated TODO.

The same one-line change in `moreh_linear_backward_multi_core_program_factory.cpp:231-235` already spells both out above its `TODO(#52269)`:

```
// Assign through the generation-neutral accessor rather than std::get<ComputeGen1Config>: the
// helper above returns whichever alternative matches `arch`, so naming Gen1 here would throw
// std::bad_variant_access on Quasar. (The local is named dfb_unpack_modes so it does not shadow
// the accessor.)
```

Worth carrying that wording to both factories here so the three converted ops read identically.

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp [hunk 2]
```diff
@@ -521,7 +521,8 @@ ttnn::device_operation::ProgramArtifacts RunningStatistics::RunningStatisticsPro
         // Re-key of the legacy unpack_to_dest_mode vector, which was indexed by CB id. The
         // writer-facing stat buffers are producer-only for this kernel, so they get no entry. An
         // omitted DFB keeps the UnpackToSrc default.
-        auto& unpack_modes = std::get<ComputeGen1Config>(compute_hw_config).unpack_modes;
+        // TODO(#52269): Quasar unpack_modes are copied from Gen1 and not yet optimized for Quasar.
+        auto& dfb_unpack_modes = unpack_modes(compute_hw_config);
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp:529]: Not for this PR, but is the rest of the sweep tracked anywhere? 17 call sites still pair `to_compute_hardware_config` with `std::get<ComputeGen1Config>`, seven of them the sibling `normalization/softmax` factories (`softmax_program_factory_general_{c_large,h_large,h_small,w_large,w_small}.cpp`, `..._attention_optimized{,_sharded}.cpp`), plus five `moreh_softmax`, three `moreh_norm/ord_other`, and `moreh_matmul_program_factory.cpp:453`.

The one I would pull forward is `ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/sdpa_program_factory.cpp:1686` — it is inside the quasar tree and keys off the live `device->arch()`, so it is the site most likely to actually reach the Gen2 alternative and throw.

#52269 tracks optimizing the Quasar values, not converting the access path, so as things stand nothing would catch the remaining sites.

**Review by Aswinmcw — APPROVED** [review]
Two notes below, neither blocking.

**Review by mcw-anasuya — APPROVED** [review]

**Review by Aswinmcw — APPROVED** [review]
Re-approving after `fc64f61`. The access-path rationale and the rename note are now recorded at both call sites, matching the wording already used in `moreh_linear_backward_multi_core_program_factory.cpp` — that was my only actionable note. Verified the accessor binds as `ComputeUnpackModes&` for either alternative, so behaviour on Gen1 is unchanged.

Still nothing tracking the ~17 remaining `to_compute_hardware_config` + `std::get<ComputeGen1Config>` pairs (#52269 covers the Quasar values, not the access path). Not for this PR.

**Comment by mcw-anasuya** [comment]
Please update deferred quasar enablement line in PR desc since it throws std::bad_variant_access

**Comment by VirdhatchaniKN** [comment]
Fair point — updated. The two were conflated: the `std::get` isn't deferred anything, it just throws. What's deferred is the Quasar values (#52269). The description now says that instead.

**Comment by VirdhatchaniKN** [comment]
Update on both threads:

@Aswinmcw — the quasar sdpa site is now #55733, fixed in #55734. Verified reachable rather than latent: `to_compute_hardware_config` returns the Gen2 config on Quasar, and the branch is gated on the caller's `fp32_dest_acc_en`, so it throws for any Quasar caller setting it. The other 16 stay op-by-op through post-port, as here — no separate tracker.

@mcw-anasuya — description reworded; it no longer calls it deferred enablement.

**Review by bbradelTT — APPROVED** [review]

**Review by mateusznowakTT — APPROVED** [review]
